envoy/network/filter.h

```diff
@@ -531,5 +531,18 @@ class MatchingData {
   virtual const ConnectionSocket& socket() const PURE;
 };
 
+/**
+ * UDP listener filter matching context data for unified matchers.
+ */
+class UdpMatchingData {
```

Comment:
I think it would be simpler if we merged this into `MatchingData` and made `socket` optional. We can modify the inputs to avoid using `socket` whenever possible, so they apply to both TCP and UDP.

Reply:
I am not sure if it works, since `MatchingData` currently have inputs like `ServerNameInput` which only works on TCP now.

Reply:
But you can have SNI in DTLS, no? So in theory, the input could work for datagrams. I'm not certain it's a good idea to combine TCP/UDP datatype, just thinking through.

Reply:
I get your idea. Do you think it would be better if we provide interfaces like `onSocket`, `onDestinationIp` which is similar to `HttpMatchingDataImpl`? @snowp

Reply:
The issue is not the callback but the way we get the local address is different between TCP, HTTP, UDP. So I guess it's fine to have separate UDPMatchingData since we have to implement Input functions three times anyways. It's unfortunate that we can't reuse code as much as we'd like.

Reply:
Yes, I mean we can make `Socket`, `DestinationIp` in `MatchingData` as `OptRef`s and use certain methods to assign their value. It is also a feasible way, but I vote for separating `MatchingData` since the way to construct a `MatchingData` with `socket` looks more natural. And I wish some day we can have template inheritance, so their will be no waste of codes.

Reply:
Given how icky the situation is without, can we take a stab at getting something generic working? I can imagine something like 

```
struct Empty {};
struct A {
    using BaseClass = Empty;
    constexpr static std::string_view name = "a";
};

struct B : A {
    using BaseClass = A;
    constexpr static std::string_view name = "b";
};

template<class T>
using BaseClass = typename T::BaseClass;

template<class T>
void traverseInheritanceChain() {
    std::cout << T::name << std::endl;

    if constexpr (!std::is_same_v<typename T::BaseClass, Empty>) {
        traverseInheritanceChain<typename T::BaseClass>();
    }
}
```
which would require all the data impls to specify either their "base class" or a singleton type. We could probably come up with something clever to avoid having to specify terminal type but might not be worth the complexity

Reply:
We are going to register `Factory<BaseMatchingData>` for `Factory<DerivedMatchingData>`, and the method we make factories to find their base class may not work, since their is no direct inheritance between `Factory<Base>` and `Factory<Derived>`.

Reply:
Thanks for digging into this, seems like it will require some more work so let's get this merged